src/programs/lsm.md

```diff
@@ -0,0 +1,191 @@
+they can decide whether a particular action is allowed to happen or not.
+
+For example, there is an LSM hook named `task_setnice` with the following
+definition:
+
+```c
+LSM_HOOK(int, 0, task_setnice, struct task_struct *p, int nice)
+```
+
+It's getting triggered when a nice value is set for any process in the system.
+If you are not familiar with the "nice value" concept, check out
+[this article](https://en.wikipedia.org/wiki/Nice_(Unix)). This LSM hook comes
+with the following arguments:
+
+* `p` is the instance of `task_struct` which represents the process on which
+  the nice value is set
+* `nice` is the nice value
+
+Developers of module/program attaching to that hook can decide whether
+particular actions of setting a nice value to a process are allowed or not.
+
+eBPF LSM programs have access to all those arguments and also one additional
+argument - `ret` - which is a return value of potential previous eBPF LSM
+programs.
+
+## Writing LSM BPF program
+
+Let's try to create an LSM eBPF program which which is triggered by
+`task_struct` hook. The purpose of this program will be denying to change a
+nice value to lower than 0 (increasing its priority) for a particular process.
+We want to see our eBPF program to make it impossible to execute the following
+command for a particular PID with a negative value:
+
+```bash
+renice [value] -p [pid]
+```
+
+eBPF projects come with two parts: eBPF program(s) and the userspace program.
+To make our example simple, we can try to deny a change of a nice value of
+the userspace process which loads the eBPF program.
+
+The first step is to create a new project:
+
+```bash
+cargo generate --name lsm-nice -d program_type=lsm -d lsm_hook=task_setnice https://github.com/aya-rs/aya-template
+```
+
+That command should create a new Aya project with an empty program attaching to
+the `task_setnice` hook. Let's go to its directory:
+
+```bash
+cd lsm-mount
+```
+
+`task_setnice` has one argument which is an instance of
+[task_struct](https://elixir.bootlin.com/linux/v5.15.3/source/include/linux/sched.h#L723),
+which is a struct defined in the Linux kernel. Therefore we need to generate a
+binding to `task_struct` with aya-gen.
+
+> If you are not familiar with aya-gen, please refer to
+> [this section](../aya/aya-gen.md).
+
+```bash
+aya-gen btf-types task_struct > lsm-nice-ebpf/src/vmlinux.rs
+```
+
+Now it's time to modify the `lsm-nice-ebpf` project and write an actual program
+there. The full program code should look like:
+
+```rust,ignore
+{{#rustdoc_include ../../examples/lsm-nice/lsm-nice-ebpf/src/main.rs:all}}
+```
+
+Let's look at the code in details.
+
+At the beginning, we include the autogenerated binding to `task_struct`:
+
+```rust,ignore
+{{#rustdoc_include ../../examples/lsm-nice/lsm-nice-ebpf/src/main.rs:vmlinux}}
+```
+
+Then we define a global variable `PID`. Here we set its value to `0`, but later
+the loader is going to set it to the actual PID.
```

Comment:
We initialize the value to `0`, but at runtime the userspace side will patch the value with the actual `pid` we're interested in. [link to set_global API]

Reply:
set_global is not in any released aya version yet, so there is no API documentation (on docs.rs)